Available dates

schafo opened the pull request:
I needed to allow several specific date ranges, so implementations support for this and figured it might be worth feeding back into the main repo.

---

**Overview of changes:**

- The structure used by `disabled-dates` has been replicated as closely as possible, to hopefully add some consistency. 

- The `isValidDisabledDates` prop validator has been reused (which might be better renamed if it's reused?).

- One difference compared to how the `disabled-dates` work is that the use of both `from` and `to` keys at the same time is not entirely supported. This seemed like a reasonable compromise given that if a specific period is disabled then the `disabled-dates` prop can be used. Open to suggestions.

- To check if `disabled-dates` or `available-dates` contain valid properties rather than just an empty object, a `checkDateObject` function has been added to the `Utils`. Perhaps this is better placed elsewhere, or the contents of the props checked in a different way?

- To ensure no breaking changes, disabled dates will take priority if both props are provided.

---

Also spotted a bug in the `isNextDisabled` and `isPrevDisabled` which has been fixed as part of this. It can easily be submitted in a separate PR if it's helpful to integrate that sooner or without integrating these changes.

I've tried to remember to run lint before committing this time 😄  but didn't appear to do much, so might be worth checking.

Let me know what your thought are on this overall @limbara.

schafo committed:
Calendar available dates, not both from/to

schafo committed:
Updated supporting files for availableDates prop

schafo committed:
Added examples for available dates

schafo committed:
Added available dates test

schafo committed:
Available dates documentation in README

schafo committed:
Clarified disabled vs available dates priority

schafo committed:
Change example dates for available dates to match disabled dates

limbara commented:
this is my opinion regarding this changes

- i think `isValidDisableDates` should be renamed, but i still haven't decided what is the name (in fact i don't know what is the right name)
- property from `disabled-dates`, `from` & `to` was never designed to work together, it's just a coincidence. i checked it just now and if the value passed to `from` is smaller than `to`, the expected result might be disabled between `from` & `to`, but now there are bugs both in calendar, prev & next button. it'll need further check if both values are provided. the same goes to `available-dates`, `from` & `to`. if `from` is smaller than `to`, the expected result the calendar dates only available between `from` & `to`. if `from` is bigger than `to` then dates is available from `from` dates, and available until `to` dates. the dates between `from` & `to` of this `available-dates` should be disabled. I don't know if it's good to have this kind of check or just leave it unrelated between `from` & `to`.
-  `checkDateObject` function name should be `isEmptyObject`( return `true` if empty ). if u really want to know the object is a date then use `isObjectDate`
-  `disabled-dates` takes priority then if `available-dates` is provided it shoud not interfere with it (both in prev-next button & calendar disabled day).

schafo committed:
isValidDisabledDate -> isValidDateRestriction

schafo committed:
checkDateObject -> isEmptyObject

schafo committed:
Support for 'to' and 'from' for availableDates

schafo commented:
Thanks for checking over and giving some input!

I've made some tweaks which hopefully address most of it:

- `isValidDisableDates` has been renamed `isValidDateRestriction`, this seemed more generic. (Feel free to change if you think there's a more suitable name.)
- I've added some extra logic/branching, which now hopefully should mean that the issue with the 'from' and 'to' being combined should work correctly for `disabledDates` and `availableDates`, in addition to when providing only 'from' or 'to'. I've also added some detail to the `availableDates` test as part of this.
- `checkDateObject` has been renamed `isEmptyObject`, and associated logic swapped
- Added some additional checking to make sure that `available-dates` does not interfere with the prev-next buttons, and an additional test case for the `availableDates` to ensure this.


While editing some of the tests for the combined use of `to` and `from` on `availableDates`, I used a `middleOfMonth` date to check that the date had [indeed been disabled](https://github.com/limbara/vue-time-date-range-picker/pull/14/commits/091f1be477c4e99100b8f3df732aedc4b798c7ab?file-filters%5B%5D=.js#diff-46a168661aa0de8f36a51d24b25bfe077068e1420c52845205f7badd88fb2d8cR44). I added this logic to `disabledDates` as well, ensuring that the date between [is indeed available](https://github.com/limbara/vue-time-date-range-picker/pull/14/commits/091f1be477c4e99100b8f3df732aedc4b798c7ab?file-filters%5B%5D=.js#diff-861b8f60fb5616bb9e31af9c94c91e80d1e42be7b5f3417ae129469397d181ccR50).

Any further thoughts are welcome!

schafo commented:
@limbara Looks like there was a build error when the CI testing ran. Are you able to retrigger it please?

limbara commented:
@schafo done. i'll be busy this whole week, so this will have to wait.

schafo commented:
Thanks! Of course - no worries. Take your time! (Just wanted to flag the error.)
